File: problems/805_split_array_with_same_average/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    bool splitArraySameAverage(vector<int>& nums) {
        int n = nums.size(), total = accumulate(nums.begin(), nums.end(), 0);
        if (n == 1) return false;
        int half = n / 2;
        vector<unordered_set<int>> dp(half + 1);
        dp[0].insert(0);
        for (int num : nums) {
            for (int k = half; k >= 1; k--) {
                for (int s : dp[k - 1]) dp[k].insert(s + num);
            }
        }
        for (int k = 1; k <= half; k++) {
            if (total * k % n == 0 && dp[k].count(total * k / n)) return true;
        }
        return false;
    }
};

File: problems/805_split_array_with_same_average/solution.py
from typing import List


class Solution:
    def splitArraySameAverage(self, nums: List[int]) -> bool:
        n = len(nums)
        total = sum(nums)
        if n == 1:
            return False
        half = n // 2
        for k in range(1, half + 1):
            if total * k % n == 0:
                break
        else:
            return False
        dp = [set() for _ in range(half + 1)]
        dp[0].add(0)
        for num in nums:
            for k in range(half, 0, -1):
                for s in dp[k - 1]:
                    dp[k].add(s + num)
        for k in range(1, half + 1):
            if total * k % n == 0 and total * k // n in dp[k]:
                return True
        return False
